### PR TITLE
Add constraint edits to fuzz tester

### PR DESCRIPTION
**Pull request by daesunp**
## Description

This PR adds constraints to edits in the fuzz tester

### CODE REVIEW

File: packages/dds/tree/src/test/shared-tree/fuzz/fuzzEditGenerators.ts [hunk 1]
```diff
@@ -427,15 +430,34 @@ export const makeTreeEditGenerator = (
 		}
 	}
 
+	const constraintGenerator = createWeightedGenerator<NodeConstraint, FuzzTestState>([
+		[
+			(state): NodeConstraint => ({
+				type: "nodeConstraint",
+				content: maybeDownPathFromNode(
+					// Selecting the parent node here, since the field is possibly empty.
+					selectTreeField(viewFromState(state), state.random, weights.fieldSelection).content
```
Comment (alex-pardes) [inline packages/dds/tree/src/test/shared-tree/fuzz/fuzzEditGenerators.ts:439]: This seems fine for now, although I do wonder if we should refactor things to have a separate `selectTreeNode` function.

File: packages/dds/tree/src/test/shared-tree/fuzz/fuzzEditGenerators.ts [hunk 2]
```diff
@@ -427,15 +430,34 @@ export const makeTreeEditGenerator = (
...
 		let change: ReturnType<typeof fieldEditChangeGenerator>;
+		let constraint: ReturnType<typeof constraintGenerator> | undefined;
 		// This could be surfaced as a config option if desired. In practice, the corresponding assert is most
 		// likely to be hit during when a test is badly configured, in which case the remedy is to fix the config,
 		// as opposed to increasing the number of attempts.
 		let attemptsRemaining = 20;
 		do {
 			fieldInfo = selectTreeField(viewFromState(state), state.random, weights.fieldSelection);
+			constraint =
+				sumWeights([weights.nodeConstraint]) > 0
```
Comment (alex-pardes) [inline packages/dds/tree/src/test/shared-tree/fuzz/fuzzEditGenerators.ts:458]:
```suggestion
				weights.nodeConstraint > 0
```

File: packages/dds/tree/src/test/shared-tree/fuzz/topLevel.fuzz.spec.ts [hunk 3]
```diff
@@ -91,6 +92,7 @@ describe("Fuzz - Top-Level", () => {
 			},
 			reconnectProbability: 0.1,
 			idCompressorFactory: deterministicIdCompressorFactory(0xdeadbeef),
+			skipMinimization: true,
```
Comment (alex-pardes) [inline packages/dds/tree/src/test/shared-tree/fuzz/topLevel.fuzz.spec.ts:95]: Why was this necessary?

File: packages/dds/tree/src/test/shared-tree/fuzz/fuzzEditReducers.ts [hunk 4]
```diff
@@ -132,6 +132,23 @@ export function applySchemaOp(state: FuzzTestState, operation: SchemaChange) {
...
+	if (fieldEdit.constraint !== undefined) {
+		switch (fieldEdit.constraint.content.type) {
+			case "nodeConstraint": {
+				const constraintNodePath = fieldEdit.constraint.content.content;
+				const constraintNode =
+					constraintNodePath !== undefined
+						? navigateToNode(tree, constraintNodePath)
+						: undefined;
+				if (constraintNode !== undefined) {
+					tree.checkout.editor.addNodeExistsConstraint(constraintNode.anchorNode);
```
Comment (alex-pardes) [inline packages/dds/tree/src/test/shared-tree/fuzz/fuzzEditReducers.ts:144]: This constraint doesn't do much if we're not in a transaction.

File: packages/dds/tree/src/test/shared-tree/fuzz/fuzzEditGenerators.ts (hunk 2 above)
Comment (daesunp) [reply packages/dds/tree/src/test/shared-tree/fuzz/fuzzEditGenerators.ts]: Refactored constraints into its own generator

File: packages/dds/tree/src/test/shared-tree/fuzz/topLevel.fuzz.spec.ts (hunk 3 above)
Comment (daesunp) [reply packages/dds/tree/src/test/shared-tree/fuzz/topLevel.fuzz.spec.ts]: Whoops, it was for debugging, but removed now

File: packages/dds/tree/src/test/shared-tree/fuzz/fuzzEditReducers.ts (hunk 4 above)
Comment (daesunp) [reply packages/dds/tree/src/test/shared-tree/fuzz/fuzzEditReducers.ts]: Refactored constraint edits to its own generator, and added acceptance condition to check if transaction is in progress

File: packages/dds/tree/src/test/shared-tree/fuzz/fuzzEditReducers.ts [hunk 5]
```diff
@@ -290,6 +294,25 @@ export function applyUndoRedoEdit(
...
+				constraintNodePath !== undefined
+					? navigateToNode(tree, constraintNodePath)
+					: undefined;
+			if (constraintNode !== undefined) {
+				tree.checkout.editor.addNodeExistsConstraint(constraintNode.anchorNode);
+			}
+			break;
+		}
+		default:
+			break;
```
Comment (alex-pardes) [inline packages/dds/tree/src/test/shared-tree/fuzz/fuzzEditReducers.ts:312]: Can we put an `unreachableCase` here, or at least a runtime assertion is we encounter an unhandled constraint type?

File: packages/dds/tree/src/test/shared-tree/fuzz/operationTypes.ts [hunk 6]
```diff
@@ -8,13 +8,27 @@ import type { DownPath } from "../../../feature-libraries/index.js";
...
 }
 
+// Currently only node constraints are supported, but more constraint types may be added in the future.
+export interface Constraint {
+	type: "constraint";
+	content: NodeConstraint;
+}
+export interface NodeConstraint {
+	type: "nodeConstraint";
+	content: undefined | DownPath;
```
Comment (alex-pardes) [inline packages/dds/tree/src/test/shared-tree/fuzz/operationTypes.ts:30]: I think a name like "path", "target", or "node" would make more sense than "content."

What does it mean if the path is undefined?

File: packages/dds/tree/src/test/shared-tree/fuzz/topLevel.fuzz.spec.ts [hunk 7]
```diff
@@ -54,6 +54,7 @@ describe("Fuzz - Top-Level", () => {
 		abort: 1,
 		fieldSelection: { optional: 1, required: 1, sequence: 3, recurse: 3 },
 		schema: 1,
+		nodeConstraint: 1,
```
Comment (alex-pardes) [inline packages/dds/tree/src/test/shared-tree/fuzz/topLevel.fuzz.spec.ts:57]: Given that constraints are conditional on being in a transaction, we might want to set the weight a bit higher (maybe 3?).

**Review by alex-pardes — APPROVED** [review]

File: packages/dds/tree/src/test/shared-tree/fuzz/fuzzEditReducers.ts (hunk 5 above)
Comment (daesunp) [reply packages/dds/tree/src/test/shared-tree/fuzz/fuzzEditReducers.ts]: Added `unreachableCase` to default case

File: packages/dds/tree/src/test/shared-tree/fuzz/topLevel.fuzz.spec.ts (hunk 7 above)
Comment (daesunp) [reply packages/dds/tree/src/test/shared-tree/fuzz/topLevel.fuzz.spec.ts]: Updated weight to 3

**Review by yann-achard-MS — APPROVED** [review]
